### PR TITLE
Consolidate file-based authorization docs

### PR DESCRIPTION
**Pull request by jhlodin**
The `hive-file-based-authorization` section should live in `/security/file-system-access-control` now that it's used by the Iceberg connector as well. We should also discuss how system-level and catalog-level access control rules interact.

### CODE REVIEW

**Review by mosabua — APPROVED** [review]
Good thing we found that duplication already existed, rather than migration the content into yet another file ;-)

**Comment by mosabua** [comment]
Also special note @kokosing .. change as requested...

**Comment by hashhar** [comment]
This is not a duplicate. Both are different mechanisms. Hive connector had it's own filed based connector access control.

**Comment by jhlodin** [comment]
@hashhar Gotcha, thanks for the insight. Just to check though, now that the Iceberg connector also supports file-based access control is there any difference between the Hive file and the Iceberg file?

Also, what actually are the differences in syntax/support between the connector and system-level file? If the syntax is the similar or the same, it would be much simpler to have a single doc on how to create the access control file and reference as needed.

**Comment by hashhar** [comment]
It's not similar at all. The connector level control allows things like column masking which the system level access control does not care about.

TL; DR: System level access control cares about what catalogs, schemas, tables and operations to allow. Connector level access control provides controls specific to the connector like being able to mask columns, expose some access mechanism present only for that connector.

IMO there is no reason to consolidate them and should be left as is.

As for whether Hive's file-based connector access control differs in any way from Iceberg's I'll defer to @findepi. But it seems Iceberg uses the system level security instead of the connector level one from Hive (from quick glance at code).

**Comment by findepi** [comment]
> As for whether Hive's file-based connector access control differs in any way from Iceberg's I'll defer to @findepi. But it seems Iceberg uses the system level security instead of the connector level one from Hive (from quick glance at code).

@hashhar see https://github.com/trinodb/trino/pull/10493 & `IcebergSecurityModule`

**Comment by hashhar** [comment]
Ah, so Iceberg's file-based security is exactly same as Hive's file-based security (which is different from system level file based access control).

**Comment by mosabua** [comment]
So the idea is that we pull this out of the Hive connector docs into some common document for whatever connector implements that .. what do you suggest we call this?

File-based access control for connectors

Also more importantly .. why do we even have that .. doesnt file system access control (which is totally a bad name btw) have all the features?
https://trino.io/docs/current/security/file-system-access-control.html

Why would any use end up using the connector specific one vs the global one? Why do we have these two implementations? There must be something .. otherwise why would we have added it to Iceberg now?

**Comment by hashhar** [comment]
> So the idea is that we pull this out of the Hive connector docs into some common document for whatever connector implements that .. what do you suggest we call this?

file-based-connector-access-control. Maybe the other should be rename to file-based-system-access-control?

> and where do we put that file .. in security or in connectors? Personally I think this should be in the security section

I'd argue for a security subpage for connectors that have ConnectorAccessControl implemented. Since these configs and docs (connector access control docs) would only apply to specific connectors and not all of them.

> Why would any use end up using the connector specific one vs the global one? Why do we have these two implementations?

The names are unfortunate.
There are two levels of access control: system and connector.
For each of those there are multiple backends/providers - one of them is file based.

The ConnectorAccessControl SPI allows you to have different implementation for access controls per-connector. i.e. if someone wanted they could write an implementation that queries the Postgres system tables and applies access control accordingly for the PostgreSQL connector, maybe queries AWS IAM and applies access control policies for ElasticSearch on AWS, or the existing SQL Standard for Hive connector which consults the Hive metastore for policies.

The SystemAccessControl on the other hand applies more uniformly to the entire system as a whole.

**Comment by jhlodin** [comment]
Noted, I have an idea of how to rework this PR to consolidate & simplify while also keeping a distinction between system and connector-level file access control. Stay tuned.

**Comment by jhlodin** [comment]
@hashhar Reworked this PR to consolidate information about file-based access control into the one doc under /security. Please take a look at this approach and see if it makes sense. I also left a couple comments on the file where we could use clarification.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 1]
```diff
@@ -1,34 +1,53 @@
...
+control where access to data and operations is defined by rules declared in
+manually-configured JSON files.
+
+There are two types of file-based access control:
+
+* **System-level access control** uses the access control plugin with a single
+  JSON file that specifies authorization rules for the whole cluster.
+* **Catalog-level access control** uses individual JSON files for each catalog
+  for granular control over the data in that catalog, including column-level
+  authorization.
```
Comment (jhlodin) [inline docs/src/main/sphinx/security/file-system-access-control.rst:15]: How do these two types of access control interact? How is precedence handled if there's a catalog file and system file that have contradicting rules?

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 2]
```diff
@@ -332,73 +365,80 @@ The example below defines the following table access policy:
...
-.. _principal_rules:
+.. _system-file-auth-principal-rules:
 
 Principal rules
 ---------------
 
 .. warning::
 
-    Principal rules are deprecated and will be removed in a future release.
+    Principal rules are deprecated and are to be removed in a future release.
```
Comment (jhlodin) [inline docs/src/main/sphinx/security/file-system-access-control.rst:441]: This note was added in release 330. Has this work been completed? Will it be completed?

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 1 above)
Comment (jhlodin) [reply docs/src/main/sphinx/security/file-system-access-control.rst:15]: Chatted with @electrum , the way this works is that `deny` rules always take precedence over `allow`. So if there is a rule in one file that conflicts with another, the operation will be denied according to either rule. As a result, there's no favoring of system vs catalog. Will add to this PR shortly.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 2 above)
Comment (hashhar) [reply docs/src/main/sphinx/security/file-system-access-control.rst:446]: No. Not removed yet.

File: docs/src/main/sphinx/release/release-330.rst [hunk 3]
```diff
@@ -39,11 +39,11 @@ Server
 Security
 --------
 
-* :ref:`principal_rules` are deprecated and will be removed in a future release.
+* :ref:`system-file-auth-principal-rules` are deprecated and will be removed in a future release.
```
Comment (hashhar) [inline docs/src/main/sphinx/release/release-330.rst:42]: IIRC changes to older release notes don't reflect on the website unless explicitly deployed by someone? @martint would know.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 4]
```diff
@@ -1,34 +1,58 @@
...
+* **System-level access control** uses the access control plugin with a single
+  JSON file that specifies authorization rules for the whole cluster.
+* **Catalog-level access control** uses individual JSON files for each catalog
+  for granular control over the data in that catalog, including column-level
+  authorization.
+
+File-based access control always prioritizes deny outcomes. If a rule in either
+a system-level or catalog-level access control file denies a particular
+operation, the operation is denied even if another rule is intended to allow
+that operation.
```
Comment (hashhar) [inline docs/src/main/sphinx/security/file-system-access-control.rst:20]: I don't think this is accurate. AFAIK it's first matching rule wins.

i.e. if I have rule_allow and rule_deny (in that order) then even in the case where both rules are possible matches only rule_allow would be applied.

@kokosing can double check me on this (I looked in `io/trino/plugin/base/security/FileBasedSystemAccessControl.java`)

**Review by hashhar — COMMENTED** [review]
Looks mostly good to me.

Can we emphasize that both types of controls have different rule syntax/grammar/structure? Easy to assume both have same structure.

File: docs/src/main/sphinx/connector/hive.rst [hunk 5]
```diff
@@ -322,7 +322,7 @@ Property Name                                      Description
 ``hive.security``                                  See :doc:`hive-security`.
 
 ``security.config-file``                           Path of config file to use when ``hive.security=file``.
-                                                   See :ref:`hive-file-based-authorization` for details.
+                                                   See :doc:`/security/file-system-access-control` for details.
```
Comment (kokosing) [inline docs/src/main/sphinx/connector/hive.rst:325]: Did you mean ``` :ref:`catalog-file-based-access-control```?

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 4 above)
Comment (kokosing) [reply docs/src/main/sphinx/security/file-system-access-control.rst:20]: That is correct. We process them from to top to bottom. If none match, then access is denied. If any match then it is allowed.

I don't recall any deny rules in file based access controls. @hashhar Can you please help to me find them? If they are there, then expected behavior is to evaluate deny rules first.

**Review by kokosing — COMMENTED** [review]
Are there any semantic changes?

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 4 above)
Comment (hashhar) [reply docs/src/main/sphinx/security/file-system-access-control.rst:20]: I couldn't find any implementations either. There is SPI for it though.

Comment (jhlodin) [reply docs/src/main/sphinx/security/file-system-access-control.rst:20]: From my conversation with @electrum last Friday it sounded like a rule_deny anywhere would override a rule_allow anywhere. If that's wrong I'll update, but want to make sure I'm not misunderstanding.

Comment (jhlodin) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Updated the introduction, please take a look.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 6]
```diff
@@ -1,34 +1,57 @@
...
+=========================
+File-based access control
+=========================
 
-This access control plugin allows you to specify authorization rules in a JSON file.
+To secure access to data in your cluster, you can implement file-based access
+control where access to data and operations is defined by rules declared in
+manually-configured JSON files.
+
+There are two types of file-based access control:
```
Comment (hashhar) [inline docs/src/main/sphinx/security/file-system-access-control.rst:9]: There are two types of access control in general (not specific to file based). Maybe we should mention this somewhere (not in this file though since it's specific to the file-based mechanisms).

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 7]
```diff
@@ -1,34 +1,57 @@
...
+
+* **System-level access control** uses the access control plugin with a single
+  JSON file that specifies authorization rules for the whole cluster.
+* **Catalog-level access control** uses individual JSON files for each catalog
+  for granular control over the data in that catalog, including column-level
+  authorization.
+
+File-based access control process from the top to bottom of the JSON file with
+an implicit deny. If a rule matches the user then the corresponding operation
+is allowed. If no rules match, the operation is denied.
```
Comment (hashhar) [inline docs/src/main/sphinx/security/file-system-access-control.rst:19]:
```suggestion
File-based access control rules are processed in order from top to bottom of the JSON rules file. The first matching rule is applied. If no rules match, the operation is denied.
```

Note that a match doesn't have to mean "allow". e.g. `"owner": false` in a rule will remove ownership for the matched user instead of granting it.

The concept is that there is a list of "rules" (i.e. what to do - could be allow/deny/whatever) and the user information is matched against the list of rules. The first rule that matches is executed. If none match then the permission is denied.

(Feel free to use this info to reword/adjust my suggestion above).

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 8]
```diff
@@ -1,34 +1,57 @@
...
 
    access-control.name=file
    security.config-file=etc/rules.json
 
-The config file is specified in JSON format. It contains rules that define
-which users have access to which resources. The rules are read from top to bottom
-and the first matching rule is applied. If no rule matches, access is denied.
+The config file is specified in JSON format. It contains rules that define which
+users have access to which resources. The rules are read from top to bottom and
+the first matching rule is applied. If no rule matches, access is denied.
```
Comment (hashhar) [inline docs/src/main/sphinx/security/file-system-access-control.rst:46]: I'd recommend to keep this paragraph and remove the one at line 17 instead. The one at line 17 is introduced prematurely and it's possible in future catalog rules and system rules will have different semantics/mechanics.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 9]
```diff
@@ -465,29 +509,180 @@ name, and allow ``alice`` and ``bob`` to use a group principal named as
...
+
+Catalog-level access control files
+==================================
+
+You can create JSON files for individual catalogs that define authorization
+rules specific to that catalog. In the catalog configuration file, you then
+add a configuration property that sets the authorization type to ``FILE`` and
+points to this configuration file. Catalog-level access control files are
+supported on a per-connector basis, refer to the connector documentation for
+more information.
```
Comment (hashhar) [inline docs/src/main/sphinx/security/file-system-access-control.rst:552]: the intro for system level control is better structured, mentioning the configs that need to be set etc.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 8 above)
Comment (jhlodin) [reply docs/src/main/sphinx/security/file-system-access-control.rst:42]: Done

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 7 above)
Comment (jhlodin) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Removed this paragraph per your comment below, since it's described in the respective intros for system-level and catalog-level.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 2 above)
Comment (jhlodin) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Got it. Removed the "will be removed later" part of this note. That sort of wording ought to be kept out of docs, because of exactly this scenario.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 9 above)
Comment (jhlodin) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: Rewrote this catalog-level intro to be more helpful and in line with the system-level intro.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 6 above)
Comment (jhlodin) [reply docs/src/main/sphinx/security/file-system-access-control.rst:9]: Currently that's described in /security/overview, which seems like the right place for it. I don't think any immediate edits need to be made there, but am open to suggestions.

Comment (hashhar) [reply docs/src/main/sphinx/security/file-system-access-control.rst:9]: /security/overview satisfies my ask. Resolved.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 10]
```diff
@@ -465,29 +504,195 @@ name, and allow ``alice`` and ``bob`` to use a group principal named as
...
+    These rules do not apply to system-defined tables in the
+    ``information_schema`` schema.
+
+Configure a catalog rules file
+------------------------------
+
+Catalog-level rules JSON files are composed of three sections, each of which is
+a list of rules that are matched in order from top to bottom. The user is
+granted the privileges from the first matching rule. All regexes default to
+``.*`` if not specified.
```
Comment (hashhar) [inline docs/src/main/sphinx/security/file-system-access-control.rst:571]:
```suggestion
The config file is specified in JSON format. It has multiple sections each of which
contains a list of rules that are matched in order from top to bottom. The user is
granted the privileges from the first matching rule. All regexes default to
``.*`` if not specified.
```

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 11]
```diff
@@ -465,29 +504,195 @@ name, and allow ``alice`` and ``bob`` to use a group principal named as
...
+* ``user`` (optional): regex to match against user name.
+* ``group`` (optional): regex to match against every user group the user belongs
+  to.
+* ``schema`` (optional): regex to match against schema name.
+* ``table`` (optional): regex to match against table name.
+* ``privileges`` (required): zero or more of ``SELECT``, ``INSERT``,
+  ``DELETE``, ``OWNERSHIP``, ``GRANT_SELECT``.
+* ``columns`` (optional): list of column constraints.
+* ``filter`` (optional): boolean filter expression for the table.
+* ``filter_environment`` (optional): environment use during filter evaluation.
```
Comment (hashhar) [inline docs/src/main/sphinx/security/file-system-access-control.rst:598]:
```suggestion
* ``filter_environment`` (optional): environment used during filter evaluation.
```

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 12]
```diff
@@ -465,29 +504,195 @@ name, and allow ``alice`` and ``bob`` to use a group principal named as
...
+
+Configure a catalog rules file
+------------------------------
+
+Catalog-level rules JSON files are composed of three sections, each of which is
+a list of rules that are matched in order from top to bottom. The user is
+granted the privileges from the first matching rule. All regexes default to
+``.*`` if not specified.
+
+Schema rules
```
Comment (hashhar) [inline docs/src/main/sphinx/security/file-system-access-control.rst:573]: None of these sections specify the name of the section to be used in the final JSON. We should fix this. No idea how it should look though.

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 13]
```diff
@@ -465,29 +504,195 @@ name, and allow ``alice`` and ``bob`` to use a group principal named as
...
+
+Column constraint
+~~~~~~~~~~~~~~~~~
+
+These constraints can be used to restrict access to column data.
+
+* ``name``: name of the column.
+* ``allow`` (optional): if false, column can not be accessed.
+* ``mask`` (optional): mask expression applied to column.
+* ``mask_environment`` (optional): environment use during mask evaluation.
```
Comment (hashhar) [inline docs/src/main/sphinx/security/file-system-access-control.rst:608]:
```suggestion
* ``mask_environment`` (optional): environment used during mask evaluation.
```

File: docs/src/main/sphinx/security/file-system-access-control.rst [hunk 14]
```diff
@@ -465,29 +504,195 @@ name, and allow ``alice`` and ``bob`` to use a group principal named as
...
+
+Filter and mask environment
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
```
Comment (hashhar) [inline docs/src/main/sphinx/security/file-system-access-control.rst:612]: Add some text to clarify this is for the value of `mask_environment` and `filter_environment`.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 11 above)
Comment (hashhar) [reply docs/src/main/sphinx/security/file-system-access-control.rst:598]: Link from here to "Filter and mask environment".

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 13 above)
Comment (hashhar) [reply docs/src/main/sphinx/security/file-system-access-control.rst:608]: Link from here to "Filter and mask environment".

**Review by hashhar — COMMENTED** [review]
Some more comments. Looks good to go otherwise.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 13 above)
Comment (jhlodin) [reply docs/src/main/sphinx/security/file-system-access-control.rst:614]: I don't think the link is needed with the restructuring/wording below

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 12 above)
Comment (jhlodin) [reply docs/src/main/sphinx/security/file-system-access-control.rst:579]: As a compromise, added some writing in the intro that shows the high level structure with the actual section names in code font.

File: docs/src/main/sphinx/security/file-system-access-control.rst (hunk 11 above)
Comment (jhlodin) [reply docs/src/main/sphinx/security/file-system-access-control.rst]: I don't think a link is needed with the restructuring/rewording below

**Review by hashhar — APPROVED** [review]

**Review by mosabua — APPROVED** [review]
Nice work. Lets ship this!

**Comment by PengCool0904** [comment]
mask_environment and  filter_environment.
The user specified for the execution environment of a specific table only declares which user is responsible for performing the filtering operation, rather than limiting the filtering operation to that user. Is that correct?
